File: benchmarks/csharp/StringChecksum.cs
using Godot;

public partial class StringChecksum : Benchmark
{

    const int ITERATIONS = 1_000_000;
    const string LOREM_IPSUM = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.";

    // Benchmark computation of checksums on a string.

    private void BenchmarkMd5BufferEmpty() {
        for (int i = 0; i < ITERATIONS; i++)
            "".Md5Buffer();
    }


    private void BenchmarkMd5BufferNonEmpty() {
        for (int i = 0; i < ITERATIONS; i++)
            LOREM_IPSUM.Md5Buffer();
    }


    private void BenchmarkSha1BufferEmpty() {
        for (int i = 0; i < ITERATIONS; i++)
            "".Sha1Buffer();
    }


    private void BenchmarkSha1BufferNonEmpty() {
        for (int i = 0; i < ITERATIONS; i++)
            LOREM_IPSUM.Sha1Buffer();
    }


    private void BenchmarkSha256BufferEmpty() {
        for (int i = 0; i < ITERATIONS; i++)
            "".Sha256Buffer();
    }


    private void BenchmarkSha256BufferNonEmpty() {
        for (int i = 0; i < ITERATIONS; i++)
            LOREM_IPSUM.Sha256Buffer();
    }


    private void BenchmarkMd5TextEmpty() {
        for (int i = 0; i < ITERATIONS; i++)
            "".Md5Text();
    }


    private void BenchmarkMd5TextNonEmpty() {
        for (int i = 0; i < ITERATIONS; i++)
            LOREM_IPSUM.Md5Text();
    }


    private void BenchmarkSha1TextEmpty() {
        for (int i = 0; i < ITERATIONS; i++)
            "".Sha1Text();
    }


    private void BenchmarkSha1TextNonEmpty() {
        for (int i = 0; i < ITERATIONS; i++)
            LOREM_IPSUM.Sha1Text();
    }


    private void BenchmarkSha256TextEmpty() {
        for (int i = 0; i < ITERATIONS; i++)
            "".Sha256Text();
    }


    private void BenchmarkSha256TextNonEmpty() {
        for (int i = 0; i < ITERATIONS; i++)
            LOREM_IPSUM.Sha256Text();
    }
}

File: gdextension/src/benchmarks/string_checksum.cpp
#include "string_checksum.h"

#include <godot_cpp/core/class_db.hpp>

using namespace godot;

void CPPBenchmarkStringChecksum::_bind_methods() {
	ClassDB::bind_method(D_METHOD("benchmark_md5_buffer_empty"), &CPPBenchmarkStringChecksum::benchmark_md5_buffer_empty);
	ClassDB::bind_method(D_METHOD("benchmark_md5_buffer_non_empty"), &CPPBenchmarkStringChecksum::benchmark_md5_buffer_non_empty);
	ClassDB::bind_method(D_METHOD("benchmark_sha1_buffer_empty"), &CPPBenchmarkStringChecksum::benchmark_sha1_buffer_empty);
	ClassDB::bind_method(D_METHOD("benchmark_sha1_buffer_non_empty"), &CPPBenchmarkStringChecksum::benchmark_sha1_buffer_non_empty);
	ClassDB::bind_method(D_METHOD("benchmark_sha256_buffer_empty"), &CPPBenchmarkStringChecksum::benchmark_sha256_buffer_empty);
	ClassDB::bind_method(D_METHOD("benchmark_sha256_buffer_non_empty"), &CPPBenchmarkStringChecksum::benchmark_sha256_buffer_non_empty);
	ClassDB::bind_method(D_METHOD("benchmark_md5_text_empty"), &CPPBenchmarkStringChecksum::benchmark_md5_text_empty);
	ClassDB::bind_method(D_METHOD("benchmark_md5_text_non_empty"), &CPPBenchmarkStringChecksum::benchmark_md5_text_non_empty);
	ClassDB::bind_method(D_METHOD("benchmark_sha1_text_empty"), &CPPBenchmarkStringChecksum::benchmark_sha1_text_empty);
	ClassDB::bind_method(D_METHOD("benchmark_sha1_text_non_empty"), &CPPBenchmarkStringChecksum::benchmark_sha1_text_non_empty);
	ClassDB::bind_method(D_METHOD("benchmark_sha256_text_empty"), &CPPBenchmarkStringChecksum::benchmark_sha256_text_empty);
	ClassDB::bind_method(D_METHOD("benchmark_sha256_text_non_empty"), &CPPBenchmarkStringChecksum::benchmark_sha256_text_non_empty);
}

void CPPBenchmarkStringChecksum::benchmark_md5_buffer_empty() {
	for (int i = 0; i < iterations; i++)
		String("").md5_buffer();
}

void CPPBenchmarkStringChecksum::benchmark_md5_buffer_non_empty() {
	for (int i = 0; i < iterations; i++)
		LOREM_IPSUM.md5_buffer();
}

void CPPBenchmarkStringChecksum::benchmark_sha1_buffer_empty() {
	for (int i = 0; i < iterations; i++)
		String("").sha1_buffer();
}

void CPPBenchmarkStringChecksum::benchmark_sha1_buffer_non_empty() {
	for (int i = 0; i < iterations; i++)
		LOREM_IPSUM.sha1_buffer();
}

void CPPBenchmarkStringChecksum::benchmark_sha256_buffer_empty() {
	for (int i = 0; i < iterations; i++)
		String("").sha256_buffer();
}

void CPPBenchmarkStringChecksum::benchmark_sha256_buffer_non_empty() {
	for (int i = 0; i < iterations; i++)
		LOREM_IPSUM.sha256_buffer();
}

void CPPBenchmarkStringChecksum::benchmark_md5_text_empty() {
	for (int i = 0; i < iterations; i++)
		String("").md5_text();
}

void CPPBenchmarkStringChecksum::benchmark_md5_text_non_empty() {
	for (int i = 0; i < iterations; i++)
		LOREM_IPSUM.md5_text();
}

void CPPBenchmarkStringChecksum::benchmark_sha1_text_empty() {
	for (int i = 0; i < iterations; i++)
		String("").sha1_text();
}

void CPPBenchmarkStringChecksum::benchmark_sha1_text_non_empty() {
	for (int i = 0; i < iterations; i++)
		LOREM_IPSUM.sha1_text();
}

void CPPBenchmarkStringChecksum::benchmark_sha256_text_empty() {
	for (int i = 0; i < iterations; i++)
		String("").sha256_text();
}

void CPPBenchmarkStringChecksum::benchmark_sha256_text_non_empty() {
	for (int i = 0; i < iterations; i++)
		LOREM_IPSUM.sha256_text();
}

CPPBenchmarkStringChecksum::CPPBenchmarkStringChecksum() {}

CPPBenchmarkStringChecksum::~CPPBenchmarkStringChecksum() {}
